src/main/java/libai/common/Pair.java

```diff
@@ -59,11 +61,33 @@ public Pair(V x, K y) {
 	 * than o.
 	 */
 	@Override
-	public int compareTo(Pair o) {
+	public int compareTo(Pair<V, K> o) {
 		return first.compareTo(o.first);
 	}
 
 	@Override
+	public int hashCode() {
+		int hash = 3;
+		hash = 53 * hash + Objects.hashCode(this.first);
+		hash = 53 * hash + Objects.hashCode(this.second);
```

Comment:
Since we are using the Objects.hashCode function already, why not go all in an do:

```
return Objects.hashCode(first, second);
```